hledger/Hledger/Cli/Commands/Run.hs

```diff
@@ -108,5 +115,46 @@ runREPL findBuiltinCommand j = do
       Just "quit" -> return ()
       Just "exit" -> return ()
       Just input -> do
-        liftIO $ runCommand findBuiltinCommand j $ parseCommand input
+        liftIO $ runCommand findBuiltinCommand $ parseCommand input
         loop
+
+{-# NOINLINE journalCache #-}
+journalCache :: MVar (Map.Map [String] Journal)
+journalCache = unsafePerformIO $ newMVar Map.empty
```

Comment:
I guess this works by running newMVar only the first time journalCache is called, relying on GHC's usual unsafePerformIO behaviour. 

Is this a common idiom ? Is it behaviour we could rely on with future GHC, or GHC compiling to wasm, or other haskell compilers like microhs ?

Are there any implications for thread safety of the Hledger.Cli library ?

Reply:
Would the uglymemo library be of any help here ? We use it for other things, eg regexps.

Reply:
MVar is used specifically for thread safety. Uglymemo does exactly this under the hood, btw: https://github.com/postgres/postgres/blob/0e42d31b0b2273c376ce9de946b59d155fac589c/src/backend/optimizer/path/costsize.c#L1171

So I am not making the situation any worse, as uglymemo is already doing all of this :)

I am not using uglymemo directly because it can't memoize `(a -> IO b)` functions, only pure ones.